### PR TITLE
Add name kwarg to Op.__call__

### LINKED ISSUE #685: Add `name` keyword argument to `Op.__call__`
### Description

Allow arbitrary `Op`s to be given names. It's a nice way to keep dprints readable and code tidy. Here's what I imagine:

Before:

```
x = pt.tensor('x', shape=(None,))
y = pt.tensor('y', shape=(None,))
z = pt.add(x, y)
z.name = 'z'
```

After:

```
x = pt.tensor('x', shape=(None,))
y = pt.tensor('y', shape=(None,))
z = pt.add(x, y, name='z')
```

### PR DESCRIPTION
**Pull request by HarshvirSandhu**
## Description
Add `name` keyword in `Op.__call__` allowing arbitrary `Op`s to be given names

## Related Issue

- [x] Closes #685 

## Checklist

- [x] Included tests that prove the fix is effective or that the new feature works

## Type of change

- [x] New feature / enhancement

### CODE REVIEW

File: pytensor/graph/op.py [hunk 1]
```diff
@@ -289,7 +289,13 @@ def __call__(self, *inputs: Any, **kwargs) -> Variable | list[Variable]:
 
         """
         return_list = kwargs.pop("return_list", False)
+        name = kwargs.pop("name", None)
         node = self.make_node(*inputs, **kwargs)
+        if isinstance(node.outputs, list):
+            if len(node.outputs) >= 1:
+                node.outputs[0].name = name
```
Comment (ricardoV94) [inline pytensor/graph/op.py:296]: Why only when there are more than one outputs and why give the name to the first output specifically?

Comment (HarshvirSandhu) [reply pytensor/graph/op.py:296]: I think I got a bit confused, it should have been something like this:
```
        name = kwargs.pop("name", None)
        if isinstance(node.outputs, list):
            for n in node.outputs:
                n.name = name
```

File: pytensor/graph/op.py [hunk 2]
```diff
@@ -289,7 +289,12 @@ def __call__(self, *inputs: Any, **kwargs) -> Variable | list[Variable]:
 
         """
         return_list = kwargs.pop("return_list", False)
+        name = kwargs.pop("name", None)
         node = self.make_node(*inputs, **kwargs)
+        print(name, kwargs)
```
Comment (Dhruvanshu-Joshi) [inline pytensor/graph/op.py:294]: Print statements are generally not expected and can be skipped.
@ricardoV94 should we make pre-commit handle this?

Comment (HarshvirSandhu) [reply pytensor/graph/op.py:294]: It seems I missed that before committing changes. Removing this line as it is not required.

**Review by twiecki — APPROVED** [review]

File: pytensor/graph/op.py [hunk 3]
```diff
@@ -289,7 +289,11 @@ def __call__(self, *inputs: Any, **kwargs) -> Variable | list[Variable]:
 
         """
         return_list = kwargs.pop("return_list", False)
+        name = kwargs.pop("name", None)
         node = self.make_node(*inputs, **kwargs)
+        if isinstance(node.outputs, list):
```
Comment (ricardoV94) [inline pytensor/graph/op.py:294]: This will always be a list. However I think we should only assign name when there's a single output or a default output

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Small tweaks/suggestion

File: pytensor/graph/op.py (hunk 3 above)
Comment (HarshvirSandhu) [reply pytensor/graph/op.py:294]: If my understanding is correct

If `node.outputs` has a length of more than 1, then they should be called `f"{name}_{i}"`
Here `i` would be the `default_output`(which is specified in Multi-output Ops)

Comment (ricardoV94) [reply pytensor/graph/op.py:294]: Not quite what I was thinking. Multi-output nodes can have a default output, which is the only thing users usually see. We should pass name directly to either to the default output of a multi-output node or the only output of a single-output node.

But for multi-output nodes without a default output we shouldn't. For those I suggest adding a numerical suffix, 0 for the first, 1 for the second, and so on

File: pytensor/graph/op.py [hunk 4]
```diff
@@ -289,7 +289,14 @@ def __call__(self, *inputs: Any, **kwargs) -> Variable | list[Variable]:
...
+            if len(node.outputs) == 1:
+                node.outputs[0].name = name
+            else:
+                for i, n in enumerate(node.outputs):
+                    n.name = f"{name}_{i}"
```
Comment (ricardoV94) [inline pytensor/graph/op.py:299]:
```suggestion
            if len(node.outputs) == 1:
                node.outputs[0].name = name
            elif self.default_output is not None:
                node.outputs[self.default_output].name = name
            else:
                for i, n in enumerate(node.outputs):
                    n.name = f"{name}_{i}"
```

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Also we need tests for multiple output with default output and not.

There are some dummy outputs we use to test basic functionality that doesn't require finding real ops.

Should be somewhere in tests/graph/....py

**Comment by ricardoV94** [comment]
Also scan seems to have a special logic for the name kwarg (see failing tests). Have to check what's going on with that

**Comment by HarshvirSandhu** [comment]
> Also scan seems to have a special logic for the name kwarg (see failing tests). Have to check what's going on with that

For scan some tests were failing because they were assigned names like `None_1`
I have put a condition to assign names only if `name` is not None

File: tests/graph/test_op.py [hunk 5]
```diff
@@ -232,3 +232,26 @@ def perform(self, *_):
...
+        def perform(self, node, inputs, outputs):
+            outputs[0] = pt.matrix()
+            outputs[1] = pt.matrix()
+
+    multi_op = MultiOutOp()
+    res = multi_op(x, name=op_name)
+    for i, r in enumerate(res):
+        assert r.name == f"{op_name}_{i}"
+
+    z = pt.add(x, y, name=op_name)
```
Comment (ricardoV94) [inline tests/graph/test_op.py:256]: Don't use pt.add. Use a single output dummy Op

File: tests/graph/test_op.py [hunk 6]
```diff
@@ -232,3 +232,26 @@ def perform(self, *_):
...
+    y = pt.vector("y")
+    op_name = "op_name"
+
+    class MultiOutOp(Op):
+        def make_node(self, *inputs):
+            outputs = [pt.dmatrix(), pt.dmatrix()]
+            return Apply(self, list(inputs), outputs)
+
+        def perform(self, node, inputs, outputs):
+            outputs[0] = pt.matrix()
```
Comment (ricardoV94) [inline tests/graph/test_op.py:248]: This is wrong but not needed. You can replace contents of perform with raise NotImplementedError

File: tests/graph/test_op.py [hunk 7]
```diff
@@ -232,3 +232,26 @@ def perform(self, *_):
...
     x = pt.TensorType(dtype="float64", shape=(1,))("x")
     assert SomeOp()(x).type == pt.dvector
+
+
+def test_op_name():
+    x = pt.vector("x")
+    y = pt.vector("y")
+    op_name = "op_name"
+
+    class MultiOutOp(Op):
```
Comment (ricardoV94) [inline tests/graph/test_op.py:242]: Also test with (and without) default output and you should have all cases covered

File: tests/graph/test_op.py [hunk 8]
```diff
@@ -232,3 +232,26 @@ def perform(self, *_):
 
     x = pt.TensorType(dtype="float64", shape=(1,))("x")
     assert SomeOp()(x).type == pt.dvector
+
+
+def test_op_name():
+    x = pt.vector("x")
```
Comment (ricardoV94) [inline tests/graph/test_op.py:238]: Use dummy test variables instead of pt.vector (the reason is this tests are in core abstract functionality). tensors ops and variables are specific implementations of these abstract objects

File: pytensor/graph/op.py [hunk 9]
```diff
@@ -289,7 +289,14 @@ def __call__(self, *inputs: Any, **kwargs) -> Variable | list[Variable]:
 
         """
         return_list = kwargs.pop("return_list", False)
+        name = kwargs.pop("name", None)
```
Comment (ricardoV94) [inline pytensor/graph/op.py:292]: Instead of popping we can make it an explicit optional kwarg in the call signature (so it's actually discoverable). Same for `return_list`. No idea why they went for this implicit approach

File: tests/graph/test_op.py [hunk 10]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+            return data
+
+        def __eq__(self, other):
+            return isinstance(other, DummyType)
+
+        def __hash__(self):
+            return hash(DummyType)
+
+        def __repr__(self):
+            return "DummyType()"
```
Comment (ricardoV94) [inline tests/graph/test_op.py:251]: Do we need a new dummy type? Can we reuse one from the existing ones for testing?

File: tests/graph/test_op.py [hunk 11]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+        def make_node(self, *inputs):
+            outputs = [dummy_variable("a")]
+            return Apply(self, list(inputs), outputs)
+
+        def perform(self, node, inputs, outputs):
+            raise NotImplementedError()
+
+    single_op = SingleOutOp()
+    res_single = single_op(x, name=op_name)
+    assert res_single.name == op_name
```
Comment (ricardoV94) [inline tests/graph/test_op.py:268]: This may be more readable. Also the name we are giving is not so much an `op_name` but a `var_name`. 
```suggestion
    res_single = single_op(x, name="test_name")
    assert res_single.name == "test_name"
```

File: tests/graph/test_op.py [hunk 12]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+
+    class MultiOutOp(Op):
+        def make_node(self, *inputs):
+            outputs = [dummy_variable("a"), dummy_variable("b")]
+            return Apply(self, list(inputs), outputs)
```
Comment (ricardoV94) [inline tests/graph/test_op.py:273]:
```suggestion
        def make_node(self, input):
            inputs = [input]
            outputs = [input.type(), input.type()]
            return Apply(self, inputs, outputs)
```

File: tests/graph/test_op.py [hunk 13]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+
+        def perform(self, node, inputs, outputs):
+            raise NotImplementedError()
+
+    multi_op = MultiOutOp()
+    res = multi_op(x, name=op_name)
+    for i, r in enumerate(res):
+        assert r.name == f"{op_name}_{i}"
```
Comment (ricardoV94) [inline tests/graph/test_op.py:281]:
```suggestion
    res = multi_op(x, name="test_name")
    for i, r in enumerate(res):
        assert r.name == f"test_name_{i}"
```

File: tests/graph/test_op.py [hunk 14]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+
+    multi_op = MultiOutOp()
+    multi_op.default_output = 1
```
Comment (ricardoV94) [inline tests/graph/test_op.py:284]: A bit of a nitpick, but more "realistic" if you make the default_output something defined when a class is initialized. Just need to define a `__init__` method that assigns it to self after calling super().__init__.

```suggestion
    multi_op = MultiOutOp(default_output=1)
```

File: tests/graph/test_op.py [hunk 15]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+
+    def dummy_variable(name):
+        return Variable(DummyType(), None, None, name=name)
+
+    x = dummy_variable("x")
+
+    class SingleOutOp(Op):
+        def make_node(self, *inputs):
+            outputs = [dummy_variable("a")]
+            return Apply(self, list(inputs), outputs)
```
Comment (ricardoV94) [inline tests/graph/test_op.py:261]: Traditionally, an Op would not assign a default name in `make_node`
  
```suggestion
        def make_node(self, input):
            inputs = [input]
            outputs = [input.type()]
            return Apply(self, inputs, outputs)
```

File: tests/graph/test_op.py (hunk 14 above)
Comment (ricardoV94) [reply tests/graph/test_op.py:284]: That way you can also parametrize the number of outputs, with a `multi_output=False|True` that reduces the lines of code in this test just a tiny bit. Then in `made_node`, you can do:

```python
if self.multi_output:
  outputs = [input.type(), input.type()]
else:
  outputs = [input.type()]
```

**Comment by ricardoV94** [comment]
Functionality looks great, just some test suggestions

File: tests/graph/test_op.py (hunk 11 above)
Comment (ricardoV94) [reply tests/graph/test_op.py:268]: Also test that by default name is None?

File: tests/graph/test_op.py [hunk 16]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
 
     x = pt.TensorType(dtype="float64", shape=(1,))("x")
     assert SomeOp()(x).type == pt.dvector
+
+
+def test_op_name():
```
Comment (ricardoV94) [inline tests/graph/test_op.py:237]:
```suggestion
def test_call_name():
```

File: tests/graph/test_op.py [hunk 17]
```diff
@@ -232,3 +232,55 @@ def perform(self, *_):
...
+
+    multi_op = MultiOutOp()
+    res = multi_op(x, name=op_name)
+    for i, r in enumerate(res):
+        assert r.name == f"{op_name}_{i}"
+
+    multi_op = MultiOutOp()
+    multi_op.default_output = 1
+    res = multi_op(x, name=op_name)
+    assert res.name == op_name
```
Comment (ricardoV94) [inline tests/graph/test_op.py:286]: Test that `res.owner.outputs` that are not the `default_output` still have `name is None`

File: tests/graph/test_op.py (hunk 10 above)
Comment (HarshvirSandhu) [reply tests/graph/test_op.py:251]: The  [existing dummy type](https://github.com/pymc-devs/pytensor/blob/main/tests/graph/test_op.py#L23-L25) takes an argument called: `thingy`

I'm not sure if that should be reused (I assume it has a specific use)

Comment (ricardoV94) [reply tests/graph/test_op.py:251]: Should be fine, it's a test Type, it takes an argument for other purposes but that shouldn't be problematic for us. Just pass None or whatever you want

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Thanks @HarshvirSandhu !
